python/Ice/cancellation/client/main.py

```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env python
+# Copyright (c) ZeroC, Inc.
+
+import Ice
+import asyncio
+import getpass
+import sys
+
+# Slice module VisitorCenter in Greeter.ice maps to Python module VisitorCenter.
+from VisitorCenter import GreeterPrx
+
+async def main():
+    # Create an Ice communicator to initialize the Ice runtime. It will be automatically destroyed at the end of the
+    # with block. We enable asyncio support by passing the current event loop to initialize.
+    with Ice.initialize(sys.argv, eventLoop=asyncio.get_running_loop()) as communicator:
+
+        # GreeterPrx is a class generated by the Slice compiler. We create a proxy from a communicator and a "stringified
+        # proxy" with the address of the target object.
+        # If you run the server on a different computer, replace localhost in the string below with the server's hostname
+        # or IP address.
+        greeter = GreeterPrx(communicator, "greeter:tcp -h localhost -p 4061")
+
+        # Create a proxy to the slow greeter. It uses the same connection as the regular greeter.
+        slowGreeter = GreeterPrx(communicator, "slowGreeter:tcp -h localhost -p 4061")
+
+        # Create another slow greeter proxy with an invocation timeout of 4 seconds (the default invocation timeout is
+        # infinite).
+        slowGreeter4s = slowGreeter.ice_invocationTimeout(4000)
+
+        # Send a request to the regular greeter and get the response.
+        greeting = await greeter.greetAsync(getpass.getuser())
+        print(greeting)
+
+
+        # Send a request to the slow greeter with the 4-second invocation timeout.
+        try:
+            greeting = await slowGreeter4s.greetAsync("alice")
+            print(f"Received unexpected greeting: {greeting}")
+        except Ice.InvocationTimeoutException as exception:
+            print(f"Caught InvocationTimeoutException, as expected: {exception}")
+
+        # Send a request to the slow greeter, and cancel this request after 4 seconds.
+        try:
+            f = slowGreeter.greetAsync("bob")
+
+            async def cancel():
```

Comment:
I would do simpler and just sleep and cancel inline. No need for a task.

Reply:
> Does Ice for Python provide any wrapper for the cancellation function that Ice C++ returns? If yes, demo that. If not, don't demo canceling an asyncio future.

Here the cancelation is forwarded from `asyncio.Future -> Ice.Future -> C++ cancel callback`. The invocation is canceled when you cancel the asyncio.Future but you don't get the InvocationCanceled becase the future is already settled as canceled.

I would remove this part is not very interesting.

Reply:
This means the demo is now purely about invocation timeout? Then, we should rename it.

Reply:
InvocationTimeout?

Reply:
Yes